File: rust/perspective/cpp/ffi.cpp
#include "ffi.h"
#include "perspective/base.h"

namespace ffi {

std::unique_ptr<Pool>
mk_pool() {
    return std::unique_ptr<Pool>(new Pool());
}

perspective::t_dtype
convert_to_dtype(DType dtype) {
    return static_cast<perspective::t_dtype>(dtype);
};

DType
get_col_dtype(const Column& col) {
    return static_cast<DType>(col.get_dtype());
}

std::shared_ptr<Table>
mk_table(rust::Vec<rust::String> column_names_ptr,
    rust::Vec<DType> data_types_ptr, std::uint32_t limit,
    rust::String index_ptr) {
    // std::shared_ptr<perspective::t_gnode> gnode;
    // gnode->init();

    std::vector<std::string> column_names;
    for (auto s : column_names_ptr) {
        std::string ss(s.begin(), s.end());
        column_names.push_back(ss);
    }
    std::vector<perspective::t_dtype> data_types;
    for (auto s : data_types_ptr) {
        data_types.push_back(convert_to_dtype(s));
    }
    std::string index(index_ptr.begin(), index_ptr.end());

    auto pool = std::make_shared<Pool>();
    auto tbl
        = std::make_shared<Table>(pool, column_names, data_types, limit, index);

    perspective::t_schema schema(column_names, data_types);
    perspective::t_data_table data_table(schema);
    data_table.init();

    // tbl->set_gnode(gnode);

    tbl->init(data_table, 0, perspective::t_op::OP_INSERT, 0);
    return tbl;
}
} // namespace ffi

File: rust/perspective/src/ffi.rs
use cxx::let_cxx_string;
use cxx::{SharedPtr, UniquePtr};

#[cxx::bridge(namespace = ffi)]
mod ffi_internal {
    #[derive(Debug)]
    enum DType {
        DTYPE_NONE,
        DTYPE_INT64,
        DTYPE_INT32,
        DTYPE_INT16,
        DTYPE_INT8,
        DTYPE_UINT64,
        DTYPE_UINT32,
        DTYPE_UINT16,
        DTYPE_UINT8,
        DTYPE_FLOAT64,
        DTYPE_FLOAT32,
        DTYPE_BOOL,
        DTYPE_TIME,
        DTYPE_DATE,
        DTYPE_ENUM,
        DTYPE_OID,
        DTYPE_OBJECT,
        DTYPE_F64PAIR,
        DTYPE_USER_FIXED,
        DTYPE_STR,
        DTYPE_USER_VLEN,
        DTYPE_LAST_VLEN,
        DTYPE_LAST,
    }

    unsafe extern "C++" {
        include!("perspective/cpp/ffi.h");
        include!("perspective/cpp/types.h");

        type Pool;
        type Table;
        type GNode;
        type DataTable;
        type Column;

        pub fn size(self: &Table) -> u32;
        pub fn get_gnode(self: &Table) -> SharedPtr<GNode>;

        pub fn get_table_sptr(self: &GNode) -> SharedPtr<DataTable>;

        pub fn get_column(self: &DataTable, name: &CxxString) -> SharedPtr<Column>;

        pub fn get_col_dtype(col: &Column) -> DType;

        pub fn mk_pool() -> UniquePtr<Pool>;

        pub fn mk_table(
            column_names: Vec<String>,
            data_types: Vec<DType>,
            limit: u32,
            index: String,
        ) -> SharedPtr<Table>;

    }
}
pub use ffi_internal::mk_pool;
pub use ffi_internal::mk_table;
pub use ffi_internal::DType;

// use crate::cpp_common::DType;

pub struct Pool {
    pool: UniquePtr<ffi_internal::Pool>,
}

impl Pool {
    pub fn new() -> Self {
        Pool {
            pool: ffi_internal::mk_pool(),
        }
    }
}

pub struct Column {
    column: SharedPtr<ffi_internal::Column>,
}

impl Column {
    pub fn get_dtype(&self) -> DType {
        ffi_internal::get_col_dtype(&self.column)
    }
}

pub struct Table {
    table: SharedPtr<ffi_internal::Table>,
}

impl Table {
    pub fn new() -> Self {
        let column_names = vec!["a".to_owned(), "psp_pkey".to_owned()];
        let data_types = vec![DType::DTYPE_INT64, DType::DTYPE_INT64];
        let limit = 100;
        let index = "a".to_string();
        let table = ffi_internal::mk_table(column_names, data_types, limit, index);
        Table { table }
    }

    pub fn get_column(&self, name: &str) -> Column {
        let gnode = self.table.get_gnode();
        let data_table = gnode.get_table_sptr();
        let_cxx_string!(n = name);
        let col = data_table.get_column(&n);
        Column { column: col }
    }

    pub fn size(&self) -> u32 {
        self.table.size()
    }
}
